### DISCUSSION TITLE
What about the filesystem? To sync or not to sync...

### DISCUSSION THREAD

**Discussion by mb21** [automerge/automerge#758] [General]
The initial [Local-first software paper](https://www.inkandswitch.com/local-first/) emphasized the role of the (local) filesystem:

> Since local-first applications store the primary copy of their data in each device’s local filesystem, the user can read and write this data anytime, even while offline. It is then synchronized with other devices sometime later, when a network connection is available.

However, the automerge library as currently implemented uses IndexedDB in the browser, and provides a binary-blob in its [`Document.save()` method](https://automerge.org/automerge-swift/documentation/automerge/document/save()).

Of course, you could argue that the browser's IndexedDB is stored somewhere on the local filesystem. But that's not really what people think of when they say "filesystem", right?

The focus on the file system intuitively feels appealing to me, as someone who trusts my machine's local filesystem and feels like that's something I understand and control. That's one of the reasons why I like plain text files (such as Markdown). So naively, a user could even wish for a plain text file representation (or at least JSON) of the automerge-document – as a sort of backup, or to mirror into a git repo.

However, as a software developer, I see the problem: any process on the local computer can potentially edit a file, even when our local-first software is not running. So how do you keep things synchronized? Do you treat the plain text files as a db-dump and overwrite any changes from other processes? Do you attempt to merge the detected changes back in, and if so do you continuously do that with a background-process, or prompt the user how to handle it on the next app startup? None of this seems exactly smooth.

So perhaps the time of the file system as a universal interface (Unix's "everything is a file") has run its course after all? Funnily, even git still uses the filesystem as its interface, where you read files from and edit things, before you commit the changes and it goes into the database that is the hidden `.git/` folder.

I'm curious about the automerge folk's thoughts on the filesystem. Has your thinking evolved on this since the initial paper, or it's just something you still want to tackle in the future?

**Comment by alexjg** [comment]
What do you mean when you say that "the automerge library as currently implemented uses IndexedDB in the browser"? Automerge itself doesn't provide any storage implementations at all, are you referring to the work in [automerge-repo](https://github.com/automerge/automerge-repo)? If so `automerge-repo` also provides a [filesystem based storage adapter](https://automerge.org/automerge-repo/modules/_automerge_automerge_repo_storage_nodefs.html) for use when you're not in the browser.

In the quote that you've referenced I don't think the filesystem is the important part - it's more that the data is on the device and doesn't require another computer to be available for the application to continue working. That said, what interests me about local first computing is that it increases user agency and I think that all else being equal, data stored on the filesystem gives the user more agency than data in `IndexedDB` because most people understand a filesystem and there's a large ecosystem of tools for working with data in the filesystem; consequently users have more ways to work with their data. 

As you note, there are security problems here, but maybe using something like the file access API in the browser would be a way to sandbox untrusted applications in the browser while also allowing trusted applications to directly use the filesystem.

**Reply by mb21** [reply]
Yes, I was indeed referring to `automerge-repo`. Thanks! I'll check out the filesystem storage adapter.

> what interests me about local first computing is that it increases user agency and I think that all else being equal, data stored on the filesystem gives the user more agency than data in IndexedDB because most people understand a filesystem and there's a large ecosystem of tools for working with data in the filesystem

Precisely. But I'm guessing the filesystem storage adapter writes a binary blob into the file? (I'll try myself later) To really profit from "a large ecosystem of tools for working with data in the filesystem", it would need to be plain text, or at least something like JSON...

> As you note, there are security problems here

I was not even thinking about security. Even if a friendly user decides to edit the file with one of the "tools for working with data in the filesystem" (say, a plain text editor), there is a synchronization issue, as the change would come in as one big chunk, not as CRDT types.

**Reply by alexjg** [reply]
The adapter does indeed just write a blob to the filesystem. I think your reference to "everything is a file" provides one answer to that. One can imagine CLI tools like `jq` which adapt other unix tools to work with automerge blobs on the local filesystem (indeed, the experimental `automerge-cli` crate in this repo implements part of that). You can already do `cat blob1 blob2 | automerge export | jq ` to get a JSON representation of an automerge encoded file (and note that the concatenation of two compressed automerge files will decode as the merge of those two files).

There is still other work to be done, it's not obvious how to edit cleanly here. One way would be to use something like [`vipe`](https://github.com/juliangruber/vipe) to edit a text field and then diff the output to generate the change in an automerge document. Maybe something like this:

```
cat doc1 doc2 | automerge edit --path 'contacts[0].name' | vipe | automerge export doc3
```

which would merge `doc1` and `doc2`, then edit the `name` field of the first element of the `contacts` array (which should be a text sequence) with the configured `EDITOR` and then save the result to `doc3`.

Just spitballing here though, I have not thought hard about a good design for this :slightly_smiling_face:

**Reply by mb21** [reply]
yeah, feels like if we want the file to really be accessible to users, then that conversion step should be done by the adapter... or perhaps things could be mounted as a virtual file-system, and the conversion step be done transparently on read...

Or perhaps Apple was right all along, and messing with plain text files on the filesystem is just "legacy" :P
